# Practical work N°1 : Recommender Systems


Auteurs : Sarra MADAD, Manelle NOUAR, Mohamed Abdelkader




# Part 1 : Manipulating Data with Pandas library

#### Create a small DataFrame small_df that only has the following features: title, release_date, budget, revenue, runtime, and genres.

In [4]:
import pandas as pd
import numpy as np

# Features :
colname = ["title", "release_date", "budget", "revenue", "runtime", "genres"]

# A small DataFrame small_df.
small_df = pd.read_csv("/home/manelle/movies_metadata.csv", usecols = colname)

small_df.head(3)

budget                                             genres release_date  \
0  30000000  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30   
1  65000000  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15   
2         0  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22   

       revenue  runtime             title  
0  373554033.0     81.0         Toy Story  
1  262797249.0    104.0           Jumanji  
2          0.0    101.0  Grumpier Old Men

#### Check out the data types of the different features. What is the type of “budget”?

In [5]:
small_df.dtypes

budget           object
genres           object
release_date     object
revenue         float64
runtime         float64
title            object
dtype: object

In [6]:
small_df['budget']

0        30000000
1        65000000
2               0
3        16000000
4               0
           ...   
45461           0
45462           0
45463           0
45464           0
45465           0
Name: budget, Length: 45466, dtype: object

On remarque que la colonne budget est de type object. Cette colonne contient des string mal formatés.

#### Convert the budget feature to float. What is the result?

In [7]:
small_df["budget"] = small_df["budget"].astype(float)
#small_df['budget'] = pd.to_numeric(small_df['budget'])

ValueError: could not convert string to float: '/ff9qCepilowshEtG2GYWwzt2bs4.jpg'

La fonction astype() ou bien to_numeric(), retourne une erreur. Je n'arrive donc pas à convertir la colonne. Il y a sûrement dans la colonne "budget" des string, qui ne peuvent pas être converti en flottant.

Autre façon de convertir la colonne, en prenant en compte les erreurs.

In [8]:
small_df["budget"] = pd.to_numeric(small_df.budget, errors='coerce')
small_df["budget"]

0        30000000.0
1        65000000.0
2               0.0
3        16000000.0
4               0.0
            ...    
45461           0.0
45462           0.0
45463           0.0
45464           0.0
45465           0.0
Name: budget, Length: 45466, dtype: float64

#### Create a function to_float(number) to convert every number field in budget to float and, if that fails, convert it to NaN.

In [27]:
def to_float(n) :
    try :
        n = float(n)
    except :
        n = np.nan
    return n

#### Apply the to_float function to all values in the budget column, then try converting to float using pandas astype. Check out your result by getting the data types for all features.

In [28]:
small_df['budget'] = small_df['budget'].apply(to_float)
small_df['budget']

0        30000000.0
1        65000000.0
2               0.0
3        16000000.0
4               0.0
            ...    
45461           0.0
45462           0.0
45463           0.0
45464           0.0
45465           0.0
Name: budget, Length: 45466, dtype: float64

On peut voir que la colonne "budget" est convertie en flottant.

#### Define a new feature called “year” that represents the year of release by using the datetime functionality from pandas as following:

In [30]:
# Convert release_date into pandas datetime format.
small_df["release_date"] = pd.to_datetime(small_df["release_date"], format='%Y-%m-%d', errors="coerce")

In [31]:
# Extract year from the datetime.
small_df["year"] = pd.DatetimeIndex(small_df["release_date"]).year

In [32]:
small_df.head(3)

budget                                             genres release_date  \
0  30000000.0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   1995-10-30   
1  65000000.0  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   1995-12-15   
2         0.0  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   1995-12-22   

       revenue  runtime             title    year  
0  373554033.0     81.0         Toy Story  1995.0  
1  262797249.0    104.0           Jumanji  1995.0  
2          0.0    101.0  Grumpier Old Men  1995.0

La colonne "year" a bien été ajoutée à notre base de donnée.

#### What are the oldest movies available in this dataset?

In [33]:
small_df.sort_values(by=['year']).head(1)

budget                               genres release_date  revenue  \
34940     0.0  [{'id': 99, 'name': 'Documentary'}]   1874-12-09      0.0   

       runtime             title    year  
34940      1.0  Passage of Venus  1874.0

Le film le plus ancien date de 1874.

#### What are the most successful (revenue feature) movies of all time ?

In [34]:
# The most successful movies of all time.
small_df.sort_values(by=['revenue'],ascending=False).head(1)

budget                                             genres  \
14551  237000000.0  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

      release_date       revenue  runtime   title    year  
14551   2009-12-10  2.787965e+09    162.0  Avatar  2009.0

Le film qui a le plus de succès est Avatar.

#### Create a new DataFrame new of only movies that earned more than $1 billion.

In [35]:
hit_movie = small_df['revenue'] > 1000000000
new = small_df[hit_movie]
new.head(3)

budget                                             genres  \
1639   200000000.0  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...   
7000    94000000.0  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
11008  200000000.0  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   

      release_date       revenue  runtime  \
1639    1997-11-18  1.845034e+09    194.0   
7000    2003-12-01  1.118889e+09    201.0   
11008   2006-06-20  1.065660e+09    151.0   

                                               title    year  
1639                                         Titanic  1997.0  
7000   The Lord of the Rings: The Return of the King  2003.0  
11008     Pirates of the Caribbean: Dead Man's Chest  2006.0

#### Create another new DataFrame new2 of only movies earned more than $1 billion, but where the budget less than $150 million.

In [36]:
reduced_budget = small_df['budget'] < 150000000
new2 = small_df[hit_movie & reduced_budget]
new2.head(3)

budget                                             genres  \
7000    94000000.0  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
17437  125000000.0  [{'id': 10751, 'name': 'Family'}, {'id': 14, '...   
30700   74000000.0  [{'id': 10751, 'name': 'Family'}, {'id': 16, '...   

      release_date       revenue  runtime  \
7000    2003-12-01  1.118889e+09    201.0   
17437   2011-07-07  1.342000e+09    130.0   
30700   2015-06-17  1.156731e+09     91.0   

                                               title    year  
7000   The Lord of the Rings: The Return of the King  2003.0  
17437   Harry Potter and the Deathly Hallows: Part 2  2011.0  
30700                                        Minions  2015.0

# Part 2: Building a simple recommender

In [37]:
data = pd.read_csv("/home/manelle/movies_metadata.csv")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date           45379 non-null  object 
 15  re

<ipython-input-37-c5c12740522e>:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("/home/manelle/movies_metadata.csv")


#### 1. Choose a metric (or score) to rate the movies on.

In [52]:
m = data["vote_count"].quantile(0.80) # On prends les 20% les mieux classés
m

50.0

Dans un premier temps, on va sélectionner les films qui sont classés dans les 20 % les mieux notés. On peut voir que 20 % des films ont reçu 50 ou plus de votes.

In [53]:
# The movie on the chart.
# Durée supérieure à 45 minutes et moins de 300 minutes.
temp = data[(data["runtime"] >= 45) & (data["runtime"] <= 300)]
temp.shape

(41930, 24)

Une autre condition préalable que nous voulons mettre en place est la durée d'exécution. Nous ne considérons que les films d'une durée supérieure à 45 minutes
et moins de 300 minutes.

#### 2. Decide on the prerequisites for the movie to be featured on the chart.

In [54]:
chart = temp[temp["vote_count"] >= m] # On garde les meilleurs scores.
chart.head(3)

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   

     original_title                                           overview  ...  \
0         Toy Story  Led by Woody, Andy's toys live happily in his ...  ...   
1           Jumanji  When siblings Judy and Peter discover an encha...  ...   
2  Grumpier Old Men  A family wedding reignites the ancient feud be...  ...   

  release_date      revenue runtime  \
0   1995-10-30  373554033.0    81.0   
1   1995-12-15  262797249.0   104.0   
2   1995-12-22          0.0   101.0   

                                    spoken_languages    status  \
0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline             title  video  \
0                                                NaN         Toy Story  False   
1          Roll the dice and unleash the excitement!           Jumanji  False   
2  Still Yelling. Still Fighting. Still Ready for...  Grumpier Old Men  False   

  vote_average vote_count  
0          7.7     5415.0  
1          6.9     2413.0  
2          6.5       92.0  

[3 rows x 24 columns]

#### 3. Calculate the score for every movie that satisfies the conditions.

In [55]:
# Calcul correspondant au Mean Rating.
C = data['vote_average'].sum()/45466 # sur le nombre d'observations
C

5.617465798618749

In [56]:
v = chart['vote_count']
v

0        5415.0
1        2413.0
2          92.0
4         173.0
5        1886.0
          ...  
45268     153.0
45339      94.0
45343     167.0
45348      55.0
45441     145.0
Name: vote_count, Length: 8963, dtype: float64

In [58]:
R = chart['vote_average']
R

0        7.7
1        6.9
2        6.5
4        5.7
5        7.7
        ... 
45268    5.3
45339    5.2
45343    4.0
45348    3.4
45441    5.2
Name: vote_average, Length: 8963, dtype: float64

In [59]:
# Calcul du Weighted Rating
WR = ((v / (v + m)) * R) + ((m /(v + m)) * C)
WR

0        7.680947
1        6.873964
2        6.189249
4        5.681495
5        7.646216
           ...   
45268    5.378194
45339    5.344953
45343    4.372688
45348    4.455936
45441    5.307043
Length: 8963, dtype: float64

#### 4. Output the list of movies in decreasing order of their scores.

In [60]:
Score = pd.DataFrame(WR,columns=["Score"])
Score # demander si c'est normal d'en avoir autant 

Score
0      7.680947
1      6.873964
2      6.189249
4      5.681495
5      7.646216
...         ...
45268  5.378194
45339  5.344953
45343  4.372688
45348  4.455936
45441  5.307043

[8963 rows x 1 columns]

In [62]:
Score['Score'].describe()

count    8963.000000
mean        6.195671
std         0.696408
min         3.158806
25%         5.731118
50%         6.180241
75%         6.648008
max         8.855096
Name: Score, dtype: float64

In [63]:
df = pd.DataFrame(data,columns=['title', 'vote_count', 'vote_average'])
df2 = Score.join(df) 
df2.sort_values(by=['Score'], ascending=False,inplace=True)
df2

Score                        title  vote_count  vote_average
10309  8.855096  Dilwale Dulhania Le Jayenge       661.0           9.1
314    8.482858     The Shawshank Redemption      8358.0           8.5
834    8.476272                The Godfather      6024.0           8.5
40251  8.366549                   Your Name.      1030.0           8.5
12481  8.289112              The Dark Knight     12269.0           8.3
...         ...                          ...         ...           ...
17708  3.577927   Birdemic: Shock and Terror        69.0           2.1
12911  3.519578               Disaster Movie       250.0           3.1
11557  3.514774                   Epic Movie       334.0           3.2
3471   3.423538            Battlefield Earth       259.0           3.0
13566  3.158806         Dragonball Evolution       475.0           2.9

[8963 rows x 4 columns]

# Part 3: Implement a user-based collaborative filtering system

In [1]:
ratings_small = pd.read_csv("/home/manelle/ratings_small.csv")
ratings_small = pd.DataFrame(ratings_small)
ratings_small

userId  movieId  rating   timestamp
0            1       31     2.5  1260759144
1            1     1029     3.0  1260759179
2            1     1061     3.0  1260759182
3            1     1129     2.0  1260759185
4            1     1172     4.0  1260759205
...        ...      ...     ...         ...
99999      671     6268     2.5  1065579370
100000     671     6269     4.0  1065149201
100001     671     6365     4.0  1070940363
100002     671     6385     2.5  1070979663
100003     671     6565     3.5  1074784724

[100004 rows x 4 columns]

1. Using the data in ratings_small.csv, construct the user rating matrix (where the rows represent the users, and the columns the movies).

In [2]:
ratings_small = ratings_small.pivot(index='userId', columns='movieId', values='rating')
ratings_small.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN     4.0     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  161084  161155  161594  161830  161918  161944  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     4.0  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     4.0  ...     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  162376  162542  162672  163949  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  
4           NaN     NaN     NaN     NaN  
5           NaN     NaN     NaN     NaN  

[5 rows x 9066 columns]

2. This matrix will be sparse (lot of unknown values), explain why. Propose a way to fill in the missing values and justify it. Implement it to have a matrix without missing values.

On peut remplacer les valeurs manquantes par 0. Mais on ne va pas utiliser cette méthode.

In [5]:
df = ratings_small.replace(np.nan, 0)
df

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0     4.0     0.0     0.0     0.0     0.0     0.0   
...         ...     ...     ...     ...     ...     ...     ...     ...   
667         0.0     0.0     0.0     0.0     0.0     4.0     0.0     0.0   
668         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
669         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
670         4.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
671         5.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

movieId  9       10      ...  161084  161155  161594  161830  161918  161944  \
userId                   ...                                                   
1           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     4.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     4.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...         ...     ...  ...     ...     ...     ...     ...     ...     ...   
667         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
668         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
669         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
670         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
671         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

movieId  162376  162542  162672  163949  
userId                                   
1           0.0     0.0     0.0     0.0  
2           0.0     0.0     0.0     0.0  
3           0.0     0.0     0.0     0.0  
4           0.0     0.0     0.0     0.0  
5           0.0     0.0     0.0     0.0  
...         ...     ...     ...     ...  
667         0.0     0.0     0.0     0.0  
668         0.0     0.0     0.0     0.0  
669         0.0     0.0     0.0     0.0  
670         0.0     0.0     0.0     0.0  
671         0.0     0.0     0.0     0.0  

[671 rows x 9066 columns]

Autre possibilité est de remplacer les valeurs par la moyenne du film. On va utiliser cette méthode.

In [6]:
df2 = ratings_small.fillna(ratings_small.mean(axis=0))
df2

movieId   1         2         3         4         5         6         7       \
userId                                                                         
1        3.87247  3.401869  3.161017  2.384615  3.267857  3.884615  3.283019   
2        3.87247  3.401869  3.161017  2.384615  3.267857  3.884615  3.283019   
3        3.87247  3.401869  3.161017  2.384615  3.267857  3.884615  3.283019   
4        3.87247  3.401869  3.161017  2.384615  3.267857  3.884615  3.283019   
5        3.87247  3.401869  4.000000  2.384615  3.267857  3.884615  3.283019   
...          ...       ...       ...       ...       ...       ...       ...   
667      3.87247  3.401869  3.161017  2.384615  3.267857  4.000000  3.283019   
668      3.87247  3.401869  3.161017  2.384615  3.267857  3.884615  3.283019   
669      3.87247  3.401869  3.161017  2.384615  3.267857  3.884615  3.283019   
670      4.00000  3.401869  3.161017  2.384615  3.267857  3.884615  3.283019   
671      5.00000  3.401869  3.161017  2.384615  3.267857  3.884615  3.283019   

movieId  8       9        10      ...  161084  161155  161594  161830  161918  \
userId                            ...                                           
1           3.8    3.15  3.45082  ...     2.5     0.5     3.0     1.0     1.5   
2           3.8    3.15  4.00000  ...     2.5     0.5     3.0     1.0     1.5   
3           3.8    3.15  3.45082  ...     2.5     0.5     3.0     1.0     1.5   
4           3.8    3.15  4.00000  ...     2.5     0.5     3.0     1.0     1.5   
5           3.8    3.15  3.45082  ...     2.5     0.5     3.0     1.0     1.5   
...         ...     ...      ...  ...     ...     ...     ...     ...     ...   
667         3.8    3.15  3.45082  ...     2.5     0.5     3.0     1.0     1.5   
668         3.8    3.15  3.45082  ...     2.5     0.5     3.0     1.0     1.5   
669         3.8    3.15  3.45082  ...     2.5     0.5     3.0     1.0     1.5   
670         3.8    3.15  3.45082  ...     2.5     0.5     3.0     1.0     1.5   
671         3.8    3.15  3.45082  ...     2.5     0.5     3.0     1.0     1.5   

movieId  161944  162376  162542  162672  163949  
userId                                           
1           5.0     4.5     5.0     3.0     5.0  
2           5.0     4.5     5.0     3.0     5.0  
3           5.0     4.5     5.0     3.0     5.0  
4           5.0     4.5     5.0     3.0     5.0  
5           5.0     4.5     5.0     3.0     5.0  
...         ...     ...     ...     ...     ...  
667         5.0     4.5     5.0     3.0     5.0  
668         5.0     4.5     5.0     3.0     5.0  
669         5.0     4.5     5.0     3.0     5.0  
670         5.0     4.5     5.0     3.0     5.0  
671         5.0     4.5     5.0     3.0     5.0  

[671 rows x 9066 columns]

Remplacer les valeurs par la moyenne des utilisateurs sur la ligne.

In [22]:
df3 = ratings_small.fillna(ratings_small.mean(axis=1))
df3

movieId  1         2         3         4       5         6         7       \
userId                                                                      
1          2.55  3.486842  3.568627  4.348039    3.91  3.261364  3.465909   
2          2.55  3.486842  3.568627  4.348039    3.91  3.261364  3.465909   
3          2.55  3.486842  3.568627  4.348039    3.91  3.261364  3.465909   
4          2.55  3.486842  3.568627  4.348039    3.91  3.261364  3.465909   
5          2.55  3.486842  4.000000  4.348039    3.91  3.261364  3.465909   
...         ...       ...       ...       ...     ...       ...       ...   
667        2.55  3.486842  3.568627  4.348039    3.91  4.000000  3.465909   
668        2.55  3.486842  3.568627  4.348039    3.91  3.261364  3.465909   
669        2.55  3.486842  3.568627  4.348039    3.91  3.261364  3.465909   
670        4.00  3.486842  3.568627  4.348039    3.91  3.261364  3.465909   
671        5.00  3.486842  3.568627  4.348039    3.91  3.261364  3.465909   

movieId    8         9         10      ...  161084  161155  161594  161830  \
userId                                 ...                                   
1        3.866379  3.755556  3.695652  ...     NaN     NaN     NaN     NaN   
2        3.866379  3.755556  4.000000  ...     NaN     NaN     NaN     NaN   
3        3.866379  3.755556  3.695652  ...     NaN     NaN     NaN     NaN   
4        3.866379  3.755556  4.000000  ...     NaN     NaN     NaN     NaN   
5        3.866379  3.755556  3.695652  ...     NaN     NaN     NaN     NaN   
...           ...       ...       ...  ...     ...     ...     ...     ...   
667      3.866379  3.755556  3.695652  ...     NaN     NaN     NaN     NaN   
668      3.866379  3.755556  3.695652  ...     NaN     NaN     NaN     NaN   
669      3.866379  3.755556  3.695652  ...     NaN     NaN     NaN     NaN   
670      3.866379  3.755556  3.695652  ...     NaN     NaN     NaN     NaN   
671      3.866379  3.755556  3.695652  ...     NaN     NaN     NaN     NaN   

movieId  161918  161944  162376  162542  162672  163949  
userId                                                   
1           NaN     NaN     NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN     NaN     NaN  
4           NaN     NaN     NaN     NaN     NaN     NaN  
5           NaN     NaN     NaN     NaN     NaN     NaN  
...         ...     ...     ...     ...     ...     ...  
667         NaN     NaN     NaN     NaN     NaN     NaN  
668         NaN     NaN     NaN     NaN     NaN     NaN  
669         NaN     NaN     NaN     NaN     NaN     NaN  
670         NaN     NaN     NaN     NaN     NaN     NaN  
671         NaN     NaN     NaN     NaN     NaN     NaN  

[671 rows x 9066 columns]

3. Calculate the cosine similarity for each pair of users, in order to obtain a user-user similarity matrix.

In [7]:
from sklearn.metrics.pairwise import cosine_similarity

cos = cosine_similarity(df2)

data = pd.DataFrame(cos)
data

0         1         2         3         4         5         6    \
0    1.000000  0.999553  0.999716  0.998778  0.999540  0.999586  0.999544   
1    0.999553  1.000000  0.999596  0.998660  0.999427  0.999456  0.999471   
2    0.999716  0.999596  1.000000  0.998832  0.999571  0.999605  0.999602   
3    0.998778  0.998660  0.998832  1.000000  0.998695  0.998725  0.998660   
4    0.999540  0.999427  0.999571  0.998695  1.000000  0.999458  0.999433   
..        ...       ...       ...       ...       ...       ...       ...   
666  0.999671  0.999476  0.999719  0.998804  0.999527  0.999575  0.999555   
667  0.999781  0.999640  0.999805  0.998907  0.999631  0.999683  0.999666   
668  0.999783  0.999650  0.999834  0.998904  0.999658  0.999684  0.999674   
669  0.999629  0.999441  0.999649  0.998766  0.999495  0.999539  0.999511   
670  0.999607  0.999491  0.999632  0.998752  0.999454  0.999545  0.999503   

          7         8         9    ...       661       662       663  \
0    0.999579  0.999740  0.999730  ...  0.999662  0.999803  0.999035   
1    0.999436  0.999595  0.999592  ...  0.999527  0.999674  0.998904   
2    0.999601  0.999746  0.999762  ...  0.999679  0.999838  0.999059   
3    0.998708  0.998843  0.998835  ...  0.998768  0.998926  0.998180   
4    0.999437  0.999588  0.999609  ...  0.999539  0.999668  0.998941   
..        ...       ...       ...  ...       ...       ...       ...   
666  0.999571  0.999718  0.999727  ...  0.999642  0.999791  0.999024   
667  0.999672  0.999831  0.999845  ...  0.999749  0.999900  0.999135   
668  0.999672  0.999825  0.999834  ...  0.999767  0.999904  0.999155   
669  0.999540  0.999667  0.999655  ...  0.999606  0.999748  0.998991   
670  0.999474  0.999643  0.999654  ...  0.999597  0.999719  0.998954   

          664       665       666       667       668       669       670  
0    0.998291  0.999631  0.999671  0.999781  0.999783  0.999629  0.999607  
1    0.998074  0.999462  0.999476  0.999640  0.999650  0.999441  0.999491  
2    0.998292  0.999728  0.999719  0.999805  0.999834  0.999649  0.999632  
3    0.997489  0.998729  0.998804  0.998907  0.998904  0.998766  0.998752  
4    0.998242  0.999471  0.999527  0.999631  0.999658  0.999495  0.999454  
..        ...       ...       ...       ...       ...       ...       ...  
666  0.998304  0.999645  1.000000  0.999780  0.999770  0.999609  0.999583  
667  0.998374  0.999727  0.999780  1.000000  0.999881  0.999737  0.999696  
668  0.998404  0.999729  0.999770  0.999881  1.000000  0.999741  0.999708  
669  0.998252  0.999550  0.999609  0.999737  0.999741  1.000000  0.999549  
670  0.998274  0.999547  0.999583  0.999696  0.999708  0.999549  1.000000  

[671 rows x 671 columns]

4. Implement a function to return the n nearest neighbors of a given user in the sense of the calculated similarity.

In [8]:
def neighbors(n,index):

    list_sorted = data[index].sort_values(ascending = False)
    return list_sorted.head(n)

In [9]:
neighbors(5,2)

userId
64     0.999693
583    0.999690
438    0.999684
335    0.999681
76     0.999678
Name: 2, dtype: float64

5. Using the preceding function, implement a function that recommends k movies—at most—for a given user. Explain how this function works.

In [ ]:
def recommended():

    return 

recommended()

6. Propose a procedure to find which value of n is appropriate to determine the recommendations for this dataset.